add `--purge`  flags to operator remove

my-git9 opened the pull request:
**Please provide a description of this PR:**

Add `--purge` flags to `istioctl operator remove` which will remove all revisions of  Istio operator.

Fix https://github.com/istio/istio/issues/41547

my-git9 committed:
add --purge flag for operator remove

Signed-off-by: xin.li <xin.li@daocloud.io>

my-git9 commented:
Some test result:
```bash
$ ./istioctl operator remove  --allrevisions

The Kubernetes version v1.18.20 is not supported by Istio unknown. The minimum supported Kubernetes version is 1.19.
Proceeding with the installation, but you might experience problems. See https://istio.io/latest/docs/setup/platform-setup/ for a list of supported versions.

All revisions of Istio operator will be removed from cluster, Proceed? (y/N) y
Removing Istio operator...
  Removed Deployment:istio-operator:istio-operator.
  Removed Deployment:istio-operator:istio-operator-v2.
  Removed Deployment:istio-operator:istio-operator-v5.
  Removed Service:istio-operator:istio-operator.
  Removed Service:istio-operator:istio-operator-v2.
  Removed Service:istio-operator:istio-operator-v5.
  Removed ServiceAccount:istio-operator:istio-operator.
  Removed ServiceAccount:istio-operator:istio-operator-v2.
  Removed ServiceAccount:istio-operator:istio-operator-v5.
  Removed ClusterRole::istio-operator.
  Removed ClusterRole::istio-operator-v2.
  Removed ClusterRole::istio-operator-v5.
  Removed ClusterRoleBinding::istio-operator.
  Removed ClusterRoleBinding::istio-operator-v2.
  Removed ClusterRoleBinding::istio-operator-v5.
✔ Removal complete
```

```bash
$ ./istioctl operator remove -r v3 --allrevisions

The Kubernetes version v1.18.20 is not supported by Istio unknown. The minimum supported Kubernetes version is 1.19.
Proceeding with the installation, but you might experience problems. See https://istio.io/latest/docs/setup/platform-setup/ for a list of supported versions.

All revisions of Istio operator will be removed, ignoring the specified revision.

All revisions of Istio operator will be removed from cluster, Proceed? (y/N) y
Removing Istio operator...
  Removed Deployment:istio-operator:istio-operator.
  Removed Deployment:istio-operator:istio-operator-v3.
  Removed Deployment:istio-operator:istio-operator-v4.
  Removed Service:istio-operator:istio-operator.
  Removed Service:istio-operator:istio-operator-v3.
  Removed Service:istio-operator:istio-operator-v4.
  Removed ServiceAccount:istio-operator:istio-operator.
  Removed ServiceAccount:istio-operator:istio-operator-v3.
  Removed ServiceAccount:istio-operator:istio-operator-v4.
  Removed ClusterRole::istio-operator.
  Removed ClusterRole::istio-operator-v3.
  Removed ClusterRole::istio-operator-v4.
  Removed ClusterRoleBinding::istio-operator.
  Removed ClusterRoleBinding::istio-operator-v3.
  Removed ClusterRoleBinding::istio-operator-v4.
✔ Removal complete
```

zirain commented:
what about `--purge` like `istioctl uninstall`?

my-git9 commented:
> istioctl uninstall

I also wanted to use `prune` before, but found that the `prune` parameter of `istioctl uninstall` deletes all resources, including the istio operator. 
At that time, I was afraid of some misunderstandings, so I chose `allversions`. Do you think we are going to use `prune` like `istioctl uninstall`?

zirain commented:
```
cmd.PersistentFlags().BoolVar(&args.purge, "purge", false, "Delete all Istio related sources for all versions")
```

prefer to use `purge` to mean `Delete all Istio operator related sources for all versions`, @istio/wg-user-experience-maintainers WDYT?

hanxiaop commented:
> ```
> cmd.PersistentFlags().BoolVar(&args.purge, "purge", false, "Delete all Istio related sources for all versions")
> ```
> 
> prefer to use `purge` to mean `Delete all Istio operator related sources for all versions`, @istio/wg-user-experience-maintainers WDYT?

@zirain I prefer to use `--purge` to delete all the resources. I think `--dry-run` flag is great to see what is going to be changed. However, it looks like there are inconsistent behaviors between `operator init` and `operator remove`. `init` only inits default operator but `remove` removes all. I think it should only remove the default operator, and let `--revision` flag remove others.

my-git9 commented:
Thanks for the suggestion, it has been modified to `--purge`, for `istioctl operator remove` not only remove the default revision, `istioctl uninstall` seems to have the same problem. Maybe we should open a separate issue to discuss, what do you think? @hanxiaop @zirain

dhawton commented on operator/cmd/mesh/operator-remove.go:
```diff
@@ -51,14 +54,20 @@ func addOperatorRemoveFlags(cmd *cobra.Command, oiArgs *operatorRemoveArgs) {
-		Args:  cobra.ExactArgs(0),
+		Args: func(cmd *cobra.Command, args []string) error {
+			if orArgs.revision == "" && !orArgs.purge {
+				return fmt.Errorf("at least one of the --revision or --purge flags must be set")
```
Without specifying a revision, should we not default to the "default" revision like elsewhere?

my-git9 replied:
Thank you for your review, it has been modified according to your suggestion. Could you review again, thanks?

ericvn commented on releasenotes/notes/41548.yaml:
```diff
@@ -0,0 +1,8 @@
+  - 41547
+releaseNotes:
+  - |
+    **Added** `--purge` flag to `istioctl operator remove` which will remove all revisions of  Istio operator.
```
Extra space
```suggestion
    **Added** `--purge` flag to `istioctl operator remove` which will remove all revisions of Istio operator.
```

my-git9 replied:
It has been updated, thanks

istio-testing commented:
@my-git9: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
gencheck_istio | 4bce4a9b54bfd27a7b7b36867c517dc13f72cee7 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio/41548/gencheck_istio/1583733203615617024) | true | `/test gencheck`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

my-git9 commented:
The test result:
```bash
$ ./istioctl operator remove

The Kubernetes version v1.18.20 is not supported by Istio unknown. The minimum supported Kubernetes version is 1.19.
Proceeding with the installation, but you might experience problems. See https://istio.io/latest/docs/setup/platform-setup/ for a list of supported versions.

Istio operator revision default will be removed from cluster, Proceed? (y/N) y
Removing Istio operator...
  Removed Deployment:istio-operator:istio-operator.
  Removed Service:istio-operator:istio-operator.
  Removed ServiceAccount:istio-operator:istio-operator.
  Removed ClusterRole::istio-operator.
  Removed ClusterRoleBinding::istio-operator.
✔ Removal complete
$
$
$./istioclt operator remove --purge

The Kubernetes version v1.18.20 is not supported by Istio unknown. The minimum supported Kubernetes version is 1.19.
Proceeding with the installation, but you might experience problems. See https://istio.io/latest/docs/setup/platform-setup/ for a list of supported versions.

All revisions of Istio operator will be removed from cluster, Proceed? (y/N) y
Removing Istio operator...
  Removed Deployment:istio-operator:istio-operator-v1.
  Removed Deployment:istio-operator:istio-operator-v2.
  Removed Service:istio-operator:istio-operator-v1.
  Removed Service:istio-operator:istio-operator-v2.
  Removed ServiceAccount:istio-operator:istio-operator-v1.
  Removed ServiceAccount:istio-operator:istio-operator-v2.
  Removed ClusterRole::istio-operator-v1.
  Removed ClusterRole::istio-operator-v2.
  Removed ClusterRoleBinding::istio-operator-v1.
  Removed ClusterRoleBinding::istio-operator-v2.
✔ Removal complete
```

zhlsunshine commented:
Hi @zirain,  in my opinion, I think user's chose should be respected. Usually user does not awareness the `--purge` flag and want to remove specification revision Istio when the revision of Istio definitely is set.

my-git9 commented:
/cc @dhawton @ericvn  @howardjohn @istio/wg-environments-maintainers

my-git9 commented:
PTAL @istio/wg-test-and-release-maintainers @istio/release-managers @istio/wg-environments-maintainers

ericvn reviewed (approved):
Test and release notes changes look good. I leave the approval of the function update to @istio/wg-environments-maintainers

litong01 commented on operator/cmd/mesh/operator-remove.go:
```diff
@@ -79,15 +82,24 @@ func operatorRemove(cmd *cobra.Command, args *RootArgs, orArgs *operatorRemoveAr
 		}
 	}
 
+	// If the user is performing purge but also specified a revision, we should warn
```
I was wondering about this case, and very much prefer this to be an error instead of proceeding. I think if both present, then an error message should be displayed and nothing should be done, considering this is a highly destructive operation.

my-git9 replied:
Updated.

litong01 replied:
@my-git9 Thanks for updating the PR.

litong01 reviewed (changes requested):
Thanks for adding this but I think when both present, it should be considered an error. And this check should be done in the validation phase.

litong01 reviewed (approved):
Thanks for your contributions.
